flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -464,7 +452,7 @@ private void pruneCheckpoints(LongPredicate pruningChecker, boolean breakOnceChe
                 long entryCheckpointId = snapshotEntry.getKey();
 
                 if (pruningChecker.test(entryCheckpointId)) {
-                    toRemove.add(snapshotEntry);
+                    toRemove.add(new SimpleEntry<>(snapshotEntry));
```

Comment:
Please add a comment why it's necessary (or even better I guess refactor the code and replace `Map.Entry` with `Tuple2` but that's probably out of scope).

Reply:
It was a bug here, `snapshotEntry` would be updated during iterating, which would cause items already added to `toRemove` to be changed. 

Nice suggestion, I have replaced `Map.Entry` with `Tuple2`.